src/replication.c

```diff
@@ -3735,32 +3764,56 @@ void replicationStartPendingFork(void) {
     if (!hasActiveChildProcess()) {
         time_t idle, max_idle = 0;
         int slaves_waiting = 0;
-        int mincapa = -1;
+        int mincapa;
+        int req;
+        int first = 1;
         listNode *ln;
         listIter li;
 
         listRewind(server.slaves,&li);
         while((ln = listNext(&li))) {
             client *slave = ln->value;
             if (slave->replstate == SLAVE_STATE_WAIT_BGSAVE_START) {
+                if (first) {
+                    /* Get first slave's requirements */
+                    req = slave->slave_req;
+                } else if (req != slave->slave_req) {
+                    /* Skip slaves that don't match */
+                    continue;
+                }
                 idle = server.unixtime - slave->lastinteraction;
                 if (idle > max_idle) max_idle = idle;
                 slaves_waiting++;
-                mincapa = (mincapa == -1) ? slave->slave_capa :
-                                            (mincapa & slave->slave_capa);
+                mincapa = first ? slave->slave_capa : (mincapa & slave->slave_capa);
+                first = 0;
             }
         }
 
         if (slaves_waiting &&
             (!server.repl_diskless_sync ||
              max_idle >= server.repl_diskless_sync_delay))
         {
-            /* Start the BGSAVE. The called function may start a
-             * BGSAVE with socket target or disk target depending on the
-             * configuration and slaves capabilities. */
-            startBgsaveForReplication(mincapa);
+            if (mincapa_out)
+                *mincapa_out = mincapa;
+            if (req_out)
+                *req_out = req;
+            return 1;
         }
     }
+
+    return 0;
+}
+
+void replicationStartPendingFork(void) {
+    int mincapa = -1;
+    int req = -1;
+
+    if (shouldStartChildReplication(&mincapa, &req)) {
+        /* Start the BGSAVE. The called function may start a
+         * BGSAVE with socket target or disk target depending on the
+         * configuration and slaves capabilities and requirements. */
+        startBgsaveForReplication(mincapa, req);
```

Comment:
not related to this line.
i think we MUST disallow any form of filter in disk-based master mode.
this will override the rdb file we use for persistence too.

This beings me to another issue, that repl-diskless-sync is disabled by default, and i'm not sure why.
maybe it's time to enable it by default, otherwise, `redis-cli --rdb-functions` will fail.

another alternative, is to make sure that rdb files that are created with a filter are never kept for persistence, don't override the persistence rdb file, and are deleted when replication is over.

Reply:
> another alternative, is to make sure that rdb files that are created with a filter are never kept for persistence, don't override the persistence rdb file, and are deleted when replication is over.

I feel this is the right direction. Do we already have a mechanism for creating RDBs not for persistence (backups, etc?). I think we need to discuss this a bit.

Reply:
currently the rename of the temp file to the permanent file is done by the child process.
we'll need it to get a different permanent file from the parent,
or better yet, just avoid rename, and the parent will find that file, use it, and delete it when it's done with it.

Reply:
Since we know we need two different file names here one for RDBs that are filtered and one for the rest, then this might be an opportunity to finally dismantle the paradigm that replication RDBs and snapshot RDBs are one and the same.
#### Pros:
- By making sure replication RDBs have a different name we can guarantee filtered ones won't be used to load from persistence.
- We can simplify the mechanism added in 5e2319c3266c6a4e0a3745c6b3a6af63dc3d9a94 and 3bab69e921039ad2e4f56c5842c4cf3e73a598dd to just always periodically delete the sync files.
- I suspect this will make the code in `syncCommand` simpler when trying to figure out if it can merge the current full sync request with previous ones or ongoing rdb forks.
#### Cons:
- There's a theoretical optimization that when we generate the sync file if snapshotting is enabled we can reduce the number of snapshots by 1 each time we generate a sync file. In practice this is very rare as snapshots are supposed to be periodic while sync files should be rarely generated.
- Currently the max disk space used up by rdb's is: old_rdb + temp_new_rdb, after such a change it'll be: old_snapshot + old_sync + temp_rdb.

@yossigo @madolson @soloestoy WDYT?

Reply:
@yoav-steinberg I don't think we should avoid using a replication RDB as a snapshot, if it's suitable (i.e. not filtered).

But there's a bigger question here - can't we assume that a function-only RDB is always diskless? The main concern around diskless replication is that in some conditions it make take a longer time and thus require more memory resources. I don't think this is really relevant for a functions-only RDB.

Reply:
one problem with that is that redis is by default configured not to allow diskless (see #9992), maybe if we change it, then it'll be less painful to just disallow filters in disk-based mode.

Reply:
@oranagra Right now it's either always disk based or always diskless, I wonder how big is the gap to make this choice ad-hoc. I guess handling `repl-diskless-sync-delay` and all kind of states where the desired / scheduled / currently running mode is incompatible will not be trivial?

Reply:
@oranagra I pushed a fix where we force filtered RDBs to be sent over sockets. It'll fail the `SYNC` command in case the slave requests a filtered RDB and doesn't have the EOF capability.

I still think we need to separate sync RDBs from snapshot RDBs, but I guess we can resolve this for now. Please review.